packages/loopring_v3/contracts/impl/libexchange/ExchangeBlocks.sol

```diff
@@ -420,6 +427,64 @@ library ExchangeBlocks
         emit BlockCommitted(S.blocks.length - 1, publicDataHash);
     }
 
+    function validateConditionalTransfers(
+        ExchangeData.State storage S,
+        uint32 newBlockIdx,
+        bytes  memory data,
+        bytes  memory auxiliaryData
+        )
+        private
+    {
+        // Read how many conditional transfers are included in the block
+        uint numConditionalTransfers;
+        assembly {
+            numConditionalTransfers := and(mload(add(data, 104)), 0xFFFFFFFF)
+        }
+        require(auxiliaryData.length == numConditionalTransfers * 6, "INVALID_AUXILIARYDATA");
+
+        // Run over all conditional transfers
+        uint numBytesPerTransfer = 13;
+        for (uint i = 0; i < auxiliaryData.length; i += 6) {
+            // auxiliaryData contains for each conditional transfer:
+            // - Transfer index: 2 bytes (the position of the conditional transfer inside data)
+            // - Salt: 4 bytes (the salt of the transfer)
+            uint index;
+            uint salt;
+            assembly {
+                index := and(mload(add(auxiliaryData, add(i, 2))), 0xFFFF)
+                salt := and(mload(add(auxiliaryData, add(i, 6))), 0xFFFFFFFF)
+            }
+            uint offset = 104 + index * numBytesPerTransfer;
+            uint transferData;
+            assembly {
+                transferData := and(mload(add(data, offset)), 0xFFFFFFFFFFFFFFFFFFFFFFFFFF)
+            }
+
+            // Check that this is a conditional transfer
+            require((transferData & 0xFF) == 1, "INVALID_AUXILIARYDATA");
+
+            // The key of the transfer is the combination of the transfer DA data and the salt
+            uint key = (transferData << 32) | salt;
+            ExchangeData.ConditionalTransferState memory transferState = S.conditionalTransfers[key];
+            require(transferState.approved, "TRANSFER_UNAUTHORIZED");
+
+            // Make sure the request was either not used yet, or used in a block that was reverted
+            if (transferState.blockIdx != 0 && transferState.blockIdx < S.blocks.length) {
```

Comment:
I'm having a hard time to understand the reason we need `ethereumBlockNumber`... 

Assuming in Etheruem block X there is a Condition tx and no virtual blocks are in block X. Then if the Condition tx got reorganized out, such a reorg it will make all the virtual blocks after block X invalid. I think the exchange relayer should handle that scenario without trying to make the previous virtual blocks valid.

In other words, if the condition for a fast withdrawal becomes invalid, the DEX's virtual blocks need to be invalidated. The operator needs to wait long enough after the condition tx to include the offchain requests.

Reply:
It wasn't added for reorgs, it was added for reverts. For reorgs we don't have to do anything onchain because if a block was reorged out it was never part of the onchain state to begin with, so it works out.

But with an actual revert where the block gets reverted by calling `revertBlock` on the exchange we do have to make sure in the smart contract that all onchain requests that were handled in blocks that were reverted can be included in a new block again. This is also the reason we store `numDepositRequestsCommitted` and `numWithdrawalRequestsCommitted` in the block. If blocks get reverted we automatically also make sure that  onchain requests need to be handled again starting at `numDepositRequestsCommitted` and `numWithdrawalRequestsCommitted` for the latest block, no matter what request were already handled in the reverted blocks. But we don't have queue for the conditional transfers so now we have to store that state for every conditional transfer independently. This is where the `ethereumBlockNumber` comes in. This allows us to detect if the block where it was included was changed because of a block revert, which means we need to allow the inclusion of the conditional transfer in a block again.